karapace/avro_compatibility.py

```diff
@@ -0,0 +1,268 @@
+from avro.schema import (
+    ARRAY, ArraySchema, BOOLEAN, BYTES, DOUBLE, ENUM, EnumSchema, Field, FIXED, FixedSchema, FLOAT, INT, LONG, MAP,
+    MapSchema, NamedSchema, NULL, RECORD, RecordSchema, Schema, STRING, UNION, UnionSchema
+)
+from copy import copy
+from enum import Enum
+from typing import cast, Dict, List, Optional
+
+
+class SchemaCompatibilityType(Enum):
+    compatible = "compatible"
+    incompatible = "incompatible"
+    recursion_in_progress = "recursion_in_progress"
+
+
+class SchemaIncompatibilityType(Enum):
+    name_mismatch = "name_mismatch"
+    fixed_size_mismatch = "fixed_size_mismatch"
+    missing_enum_symbols = "missing_enum_symbols"
+    reader_field_missing_default_value = "reader_field_missing_default_value"
+    type_mismatch = "type_mismatch"
+    missing_union_branch = "missing_union_branch"
+
+
+class AvroRuntimeException(Exception):
+    pass
+
+
+class SchemaCompatibilityResult:
+    def __init__(
+        self,
+        compatibility: SchemaCompatibilityType = SchemaCompatibilityType.recursion_in_progress,
+        incompatibilities: List[SchemaIncompatibilityType] = None,
+        message: str = "",
+    ):
+        self.message = message
+        self.compatibility = compatibility
+        self.incompatibilities = incompatibilities or []
+
+    def merged_with(self, that):
+        that = cast(SchemaCompatibilityResult, that)
+        merged = copy(self.incompatibilities)
+        merged.extend(copy(that.incompatibilities))
+        if self.compatibility is SchemaCompatibilityType.compatible:
+            compat = that.compatibility
+        else:
+            compat = self.compatibility
+        return SchemaCompatibilityResult(compat, merged)
+
+    @staticmethod
+    def compatible():
+        return SchemaCompatibilityResult(SchemaCompatibilityType.compatible)
+
+    @staticmethod
+    def incompatible(incompat_type: SchemaIncompatibilityType, message: str, location: List[str]):
+        path = "/".join(location)
+        message = f"{path}: {message}"
+
+        ret = SchemaCompatibilityResult(SchemaCompatibilityType.incompatible, [incompat_type])
+        ret.message = message
+        return ret
+
+    def __str__(self):
+        return f"{self.compatibility}: {self.message}"
+
+
+class ReaderWriter:
+    def __init__(self, reader: Schema, writer: Schema) -> None:
+        self.reader, self.writer = reader, writer
+
+    def __hash__(self) -> int:
+        return id(self.reader) ^ id(self.writer)
```

Comment:
I think this is incorrect. `hash(a)` should be equal `hash(b)` if `a` equals `b`. However, `id` won't satisfy this property. Try e.g. this:

```python
schema = json.dumps({"name": "foo", "type": "record", "fields": [{"type": "int", "name": "f1"}]})

rw1 = ReaderWriter(Parse(schema), Parse(schema))
rw2 = ReaderWriter(Parse(schema), Parse(schema))

print(rw1 == rw2)
print(hash(rw1) == hash(rw2))
```

```
True
False
```

Reply:
`hash((self.reader, self.writer))` might have been a way, but `Schema` is not hashable.

Reply:
I lifted the code from these lines https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/SchemaCompatibility.java#L163-L166. My java is a bit rusty, but i figured the closest i can get to `System.identityHashCode` is by using the `id` function

Reply:
Aha, the trick is here: https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/SchemaCompatibility.java#L175

They use pointer comparison in `equals`. In this case `identityHashCode` makes sense and doesn't violate the `hashCode` - `equals` properties. (Though I don't immediately understand why they need this whole thing instead of normal equality.)

To follow this, we need to use `is` instead of `==` in `__eq__`.

Reply:
fixed